src/ClientTransactionTask.cc

```diff
@@ -0,0 +1,562 @@
+{
+}
+
+/**
+ * Find and return the cache entry identified by the given key.
+ *
+ * \param key
+ *      Key of the object contained in the cache entry that should be returned.
+ * \return
+ *      Returns a pointer to the cache entry if found.  Returns NULL otherwise.
+ *      Pointer is invalid once the commitCache is modified.
+ */
+ClientTransactionTask::CacheEntry*
+ClientTransactionTask::findCacheEntry(Key& key)
+{
+    CacheKey cacheKey = {key.getTableId(), key.getHash()};
+    CommitCacheMap::iterator it = commitCache.lower_bound(cacheKey);
+    CacheEntry* entry = NULL;
+
+    while (it != commitCache.end()) {
+        if (cacheKey < it->first) {
+            break;
+        } else if (it->second.objectBuf) {
+            Key otherKey(it->first.tableId,
+                         it->second.objectBuf->getKey(),
+                         it->second.objectBuf->getKeyLength());
+            if (key == otherKey) {
+                entry = &it->second;
+                break;
+            }
+        }
+        it++;
+    }
+    return entry;
+}
+
+/**
+ * Return the transaction commit decision if a decision has been reached.
+ * Otherwise, INVALID will be returned.
+ */
+WireFormat::TxDecision::Decision
+ClientTransactionTask::getDecision()
+{
+    if (state != DECISION && state != DONE)
+        return WireFormat::TxDecision::INVALID;
+    return decision;
+}
+
+/**
+ * Inserts a new cache entry with the provided key and value.  Other members
+ * of the cache entry are left to their default values.  This method must not
+ * be called once the transaction has started committing.
+ *
+ * \param tableId
+ *      The table containing the desired object (return value from
+ *      a previous call to getTableId).
+ * \param key
+ *      Variable length key that uniquely identifies the object within tableId.
+ *      It does not necessarily have to be null terminated.
+ * \param keyLength
+ *      Size in bytes of the key.
+ * \param buf
+ *      Address of the first byte of the new contents for the object;
+ *      must contain at least length bytes.
+ * \param length
+ *      Size in bytes of the new contents for the object.
+ * \return
+ *      Returns a pointer to the inserted cache entry.  Pointer is invalid
+ *      once the commitCache is modified.
+ */
+ClientTransactionTask::CacheEntry*
+ClientTransactionTask::insertCacheEntry(uint64_t tableId, const void* key,
+        uint16_t keyLength, const void* buf, uint32_t length)
+{
+    Key keyObj(tableId, key, keyLength);
+    CacheKey cacheKey = {keyObj.getTableId(), keyObj.getHash()};
+    CommitCacheMap::iterator it = commitCache.insert(
+            std::make_pair(cacheKey, CacheEntry()));
+    it->second.objectBuf = new ObjectBuffer();
+    Object::appendKeysAndValueToBuffer(
+            keyObj, buf, length, it->second.objectBuf, true);
+    return &it->second;
+}
+
+/**
+ * Make incremental progress toward committing the transaction.
+ */
+void
+ClientTransactionTask::performTask()
```

Comment:
Overall, the code for this rules engine is quite clean; nice job.

My only complaint is that I wonder if there's some way to reduce the duplication of code between the 2 RPC types.